### DISCUSSION TITLE
[Q&A] How to view Slint Preview on Zed editor

### DISCUSSION THREAD

**Discussion by 0xkelvin** [slint-ui/slint#6762] [Q&A]
<img width="750" alt="image" src="https://github.com/user-attachments/assets/8754841a-a0e3-4102-9bd5-38c156ad4622">

Do we have show preview feature on Zed editor ?

**Comment by ogoffart** [comment]
I am not familiar with the zed editor, but you should look for "code actions" or "code lenses" when having the cursor on the name of a component in a .slint file.

(I'm assuming you have the [extension](https://github.com/slint-ui/slint/tree/master/editors#zed) )

If you figure it out, please leave a comment on how to do it.

**Comment by SO9010** [comment]
@0xkelvin I had a look at the code for the Zed extension, it looks like it's mostly just highlighting for now:

https://gitlab.com/flukejones/zed-slint

I am assuming that you are using the slint extension from the Zed extension store!

**Comment by srtnnm** [comment]
> @0xkelvin I had a look at the code for the Zed extension, it looks like it's mostly just highlighting for now:
> 
> https://gitlab.com/flukejones/zed-slint
> 
> I am assuming that you are using the slint extension from the Zed extension store!

it seems that zed-editor doesn't give plugins as many features as vscode does

**Comment by hunger** [comment]
The zed devs I spoke with at rustlab said that Zed does not yet support custom commands.

Showing the preview is a custom command, so I think that explains why this does not work.

**Reply by szecket** [reply]
I just started using the Zed editor with slint and have been using the `slint-viewer`

Install using `cargo install slint-viewer`

I can then load a specific component from a specific file

**Reply by whatakey** [reply]
> The zed devs I spoke with at rustlab said that Zed does not yet support custom commands.

Hello @hunger 

By custom commands do you means [Tasks](https://zed.dev/docs/tasks) and [Runnables](https://zed.dev/blog/zed-decoded-tasks#runnables)? if so, both are on stable Zed since May 2024

**Reply by hunger** [reply]
No, I mean custom `Command`s executed by the language server via [`workspace/executeCommand`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_executeCommand).

Our LS defines "Show Preview" as such a command and tells the editor to put that in all the places it makes sense. When the editor triggers the `Show Preview` command, then the langauge server opens the live-preview UI built into itself.

**Comment by nic-obert** [comment]
If anyone is willing and able to implement the preview feature for Slint files, go ahead and make a pull request to the [zed-slint](https://gitlab.com/flukejones/zed-slint/) project. [An issue](https://gitlab.com/flukejones/zed-slint/-/issues/3) proposing this feature has already been filed there a couple of weeks ago, but the repository owner hasn't responded yet.

**Reply by hunger** [reply]
It would be cool if we could get the live-preview and not just the slint-viewer... The live-preview is built into the slint LS, it works in every editor that actually supports the LS defining commands and code actions.

From what I understood in Nov. when meeting zed devs was that Zed did not have the "call commands defined by the LS" feature. Once it is there, it should be pretty easy to make this work in Zed as well.

**Comment by armyofevilrobots** [comment]
It's been a minute, but I just tested the live preview in Zed. You can right click on the name of a component, ie "MainWindow" in this example:

```rust
export component MainWindow inherits Window {
```

You'll see the "code actions" option, which will bring up the preview option. This option will bring up the preview window.

**Reply by karmakarmeghdip** [reply]
You can also press Ctrl + . to open the code actions menu and choose show preview. I have been using Zed for Slint and Rust and it's pretty flawless

**Reply by megabyte6** [reply]
@armyofevilrobots @karmakarmeghdip how are you guys getting it to work? I just get "Error: resolving a code action" when I try

**Reply by karmakarmeghdip** [reply]
> @armyofevilrobots @karmakarmeghdip how are you guys getting it to work? I just get "Error: resolving a code action" when I try

You might be having an issue with your lsp, try installing the slint-lsp manually and putting it in your path and change your zed slint extension settings to use the lsp from your path

**Reply by megabyte6** [reply]
> > @armyofevilrobots @karmakarmeghdip how are you guys getting it to work? I just get "Error: resolving a code action" when I try
> 
> You might be having an issue with your lsp, try installing the slint-lsp manually and putting it in your path and change your zed slint extension settings to use the lsp from your path

@karmakarmeghdip I missed your reply! I never actually tried the slint-lsp separately. I didn't realize it was even a project. I'm assuming you meant the one on [crates.io](https://crates.io/crates/slint-lsp). At some point, the zed team must have fixed it, cause it just works now! Thanks for your help!

(please @ me if you need a response. thank you!)
